Week4/HW04.hs

```diff
@@ -0,0 +1,149 @@
+    -- because its taking extra storage to save the new list 
+    -- without trailing zeros and then does the matching
+    P a == P b                      = withoutTrailingZerosA == withoutTrailingZerosB
+        where 
+            withoutTrailingZerosA   = dropWhileEnd (== 0) a 
+            withoutTrailingZerosB   =  dropWhileEnd (== 0) b
+        
+
+-- Exercise 3 -----------------------------------------
+
+-- e == exponent 
+-- c == count of the polynomial term, is it first, second, third and so on
+showHelper :: (Num a, Show a, Eq a) => Poly a -> Int -> Int -> String
+showHelper (P (p:ps)) e c
+    | p == 0                = showHelper (P ps) (e-1) (c+1)
+    | e == 0                = leadingPlus ++ show p 
+    -- I was trying to get nested guards here instead of using [if-then-else] or [case] but wasnt able to 
+    | e == 1                = leadingPlus ++ leadingNum ++ "x"  ++ showHelper (P ps) (e-1) (c+1)
+    | otherwise             = leadingPlus ++ leadingNum ++ "x^" ++ show e ++ showHelper (P ps) (e-1) (c+1)
+    where 
+        leadingPlus         = case c of 
+                0           -> "" 
+                _           -> " + "
+        leadingNum          = case p of 
+                1           -> ""
+                anyP        -> show anyP
+
+
+showHelper (P []) _ _       = ""
+
+instance (Num a, Eq a, Show a) => Show (Poly a) where
+    show (P p) = 
+        showHelper (P $ reverse withoutTrailingZeros ) (length withoutTrailingZeros - 1) 0
+            where 
+                withoutTrailingZeros = dropWhileEnd (==0) p
+
+-- Exercise 4 -----------------------------------------
+
+plus :: Num a               => Poly a -> Poly a -> Poly a
+plus (P polyA) (P polyB)    = P $ zipWith (+) (polyA ++ replicate leadingZerosToA 0) (polyB ++ replicate leadingZerosToB 0)
+    where 
+        lA                  = length polyA 
+        lB                  = length polyB
+        leadingZerosToA 
+            | lA < lB       = lB - lA
+            | otherwise     = 0
+        leadingZerosToB 
+            | lB < lA       = lA - lB
+            | otherwise     = 0
+
+-- Exercise 5 -----------------------------------------
+
+-- c is the coeff being multiplied from the first polynomial
+-- e is the exponentioal term being used for multiplication from first polynomial
+
+-- timesHelper (P [1, 4, 2]) 1 0 == P [1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 1 1 == P [0, 1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 1 2 == P [0, 0, 1, 4, 2]
+-- timesHelper (P [1, 4, 2]) 2 2 == P [0, 0, 2, 8, 4]
+
+-- 1. add leading zeros list of size == e 
+-- and multiply rest with c
+timesHelper ::  (Num a)     => Poly a -> a -> Int -> Poly a
+timesHelper (P p) c e       = P $ replicate e 0 ++ map (*c) p
+
+-- getAllTimes 0 (P [1, 1, 2]) (P [1, 4, 2]) == [P [1, 4, 2], P [0, 1, 4, 2], P [0, 0, 2, 8, 4]]
+getAllTimes ::  (Num a)     => Int -> Poly a -> Poly a -> [Poly a]
+getAllTimes _ (P []) _      = []
+getAllTimes n (P [p]) q     = [timesHelper q p n]
```

Comment:
You shouldn't need this middle equation, because it's a special case of the last equation.

Reply:
Sorry but I didnt understand this and the review below and how they are a special case for last question, could you pls tell which question are you referring to ?

Reply:
I mean the middle equation in 
```haskell
getAllTimes _ (P []) _      = []
getAllTimes n (P [p]) q     = [timesHelper q p n]
getAllTimes n (P (p:ps)) q  = timesHelper q p n :  getAllTimes (n+1) (P ps) q
```
can be dropped because `[p]` is just a special case of `(p:ps)` with `ps == []`

So this definition can be simplified to
```haskell
getAllTimes _ (P []) _      = []
getAllTimes n (P (p:ps)) q  = timesHelper q p n :  getAllTimes (n+1) (P ps) q
```

Reply:
And regarding the following comment you already simplified it greatly by using `foldr (+) (P [])` which is even better than what I suggested.

Reply:
got it now, thanks a lot.